413 Request Entity Too Large on -full-geometry import

Hey there 👋 

I'm trying out photon locally right now with an extract of Germany. For this I wanted to give #823 a spin and import using:

```
... -nominatim-import 
       -host nominatim 
       -port 5432 
       -database nominatim 
       -user nominatim 
       -password topsecret 
       -full-geometry
```

Some time into the import to the embedded photon instance I get an `HTTP 413 Request Entity Too Large` which to my novice eye suggests that the bulk size of 10_000 (see [opensearch/Importer.java#L29](https://github.com/komoot/photon/blob/c6dc628afc516c3258afe23687bf79312d9743ed/src/main/java/de/komoot/photon/opensearch/Importer.java#L29)) could use some tweaking now that a document can be significantly larger than before. Would you come to the same conclusion, and would you accept a PR to make the batch size in [opensearch/Importer.java#L29](https://github.com/komoot/photon/blob/c6dc628afc516c3258afe23687bf79312d9743ed/src/main/java/de/komoot/photon/opensearch/Importer.java#L29) configurable? Maybe via an import configuration in [here](https://github.com/komoot/photon/blob/c6dc628afc516c3258afe23687bf79312d9743ed/src/main/java/de/komoot/photon/cli/CommandImport.java#L18)? 

### stack trace
```
photon-import-1  | [2026-06-19T22:48:11,213][INFO ][d.k.p.n.ImportThread     ] Imported 27300000 documents [22769.391932128863/second]
photon-import-1  | [2026-06-19T22:48:14,096][INFO ][d.k.p.n.m.NominatimAddressCache] Loaded 14 address places for country dk
photon-import-1  | [2026-06-19T22:48:14,490][INFO ][d.k.p.n.m.NominatimAddressCache] Loaded 370 address places for country fr
photon-import-1  | [2026-06-19T22:48:15,054][INFO ][d.k.p.n.m.NominatimAddressCache] Loaded 58 address places for country lu
photon-import-1  | [2026-06-19T22:48:15,221][ERROR][d.k.p.n.ImportThread     ] Import error.
photon-import-1  | java.lang.RuntimeException: Error inserting new documents
photon-import-1  | 	at de.komoot.photon.opensearch.Importer.add(Importer.java:79)
photon-import-1  | 	at de.komoot.photon.nominatim.ImportThread.addDocument(ImportThread.java:48)
photon-import-1  | 	at de.komoot.photon.nominatim.NominatimImporter.lambda$readCountry$1(NominatimImporter.java:106)
photon-import-1  | 	at org.springframework.jdbc.core.JdbcTemplate$RowCallbackHandlerResultSetExtractor.extractData(JdbcTemplate.java:1764)
photon-import-1  | 	at org.springframework.jdbc.core.JdbcTemplate.lambda$query$0(JdbcTemplate.java:739)
photon-import-1  | 	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:670)
photon-import-1  | 	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:732)
photon-import-1  | 	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:757)
photon-import-1  | 	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:783)
photon-import-1  | 	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:788)
photon-import-1  | 	at de.komoot.photon.nominatim.NominatimImporter.readCountry(NominatimImporter.java:82)
photon-import-1  | 	at de.komoot.photon.App.importFromDatabase(App.java:214)
photon-import-1  | 	at de.komoot.photon.App.startNominatimImport(App.java:165)
photon-import-1  | 	at de.komoot.photon.App.runPhoton(App.java:88)
photon-import-1  | 	at de.komoot.photon.App.main(App.java:49)
photon-import-1  | Caused by: java.lang.RuntimeException: Bulk request failed
photon-import-1  | 	at de.komoot.photon.opensearch.BulkRetryHelper.sendWithRetry(BulkRetryHelper.java:61)
photon-import-1  | 	at de.komoot.photon.opensearch.Importer.submitLoop(Importer.java:182)
photon-import-1  | 	at java.base/java.lang.Thread.run(Unknown Source)
photon-import-1  | Caused by: org.opensearch.client.transport.httpclient5.ResponseException: method [POST], host [http://127.0.0.1:9201], URI [/_bulk], status line [HTTP/1.1 413 Request Entity Too Large]
```

--- a/src/main/java/de/komoot/photon/opensearch/Importer.java
+++ b/src/main/java/de/komoot/photon/opensearch/Importer.java
@@ -25,8 +25,27 @@ import java.util.concurrent.atomic.AtomicReference;
 public class Importer implements de.komoot.photon.Importer {
     private static final Logger LOGGER = LogManager.getLogger();
 
-    /** Documents per bulk request. */
+    /** Upper bound on documents per bulk request. A bulk may flush earlier once {@link #MAX_BULK_BYTES} is reached. */
     private static final int BULK_SIZE = 10000;
+
+    /**
+     * Soft cap on the estimated serialized size of a bulk request body, in bytes. With
+     * {@code -full-geometry} a single document can carry a large polygon, so the document
+     * count alone no longer bounds the request size. Kept well below OpenSearch's default
+     * {@code http.max_content_length} (100 MB) to leave room for estimation slack and the
+     * per-operation action metadata that NDJSON adds on top of the documents.
+     */
+    private static final long MAX_BULK_BYTES = 50L * 1024 * 1024;
+
+    /**
+     * Estimated bytes per geometry coordinate in the GeoJSON encoding (two formatted doubles
+     * plus brackets and separators). Deliberately generous so the byte estimate runs high and
+     * we flush early rather than risk a 413.
+     */
+    static final int BYTES_PER_GEOMETRY_POINT = 40;
+
+    /** Generous flat allowance for a document's non-geometry fields plus its bulk action line. */
+    static final int BASE_DOC_BYTES = 1024;
 
     /** Fallback refresh_interval if the original value cannot be read. Matches the OpenSearch default. */
     private static final String FALLBACK_REFRESH_INTERVAL = "1s";
@@ -52,6 +71,7 @@ public class Importer implements de.komoot.photon.Importer {
     private static final class Bulk {
         BulkRequest.Builder builder = new BulkRequest.Builder();
         int todo = 0;
+        long bytes = 0;
     }
 
     public Importer(OpenSearchClient client, int maxConcurrentRequests, boolean tuneRefresh) {
@@ -96,10 +116,26 @@ public class Importer implements de.komoot.photon.Importer {
                 final String uuid = PhotonDoc.makeUid(placeID, objectId++);
                 b.builder.operations(op -> op.create(i -> i.index(PhotonIndex.NAME).id(uuid).document(doc)));
             }
-            if (++b.todo >= BULK_SIZE) {
+            b.bytes += estimateSize(doc);
+            // A single doc over MAX_BULK_BYTES still flushes on its own; an oversized request
+            // can't be split any smaller anyway.
+            if (++b.todo >= BULK_SIZE || b.bytes >= MAX_BULK_BYTES) {
                 flush(b);
             }
         }
+    }
+
+    /**
+     * Rough estimate of a document's serialized size, dominated by its full geometry. Used only
+     * to decide when to flush a bulk; it errs on the high side so we stay under the HTTP body limit.
+     */
+    static long estimateSize(PhotonDoc doc) {
+        long size = BASE_DOC_BYTES;
+        final var geom = doc.getGeometry();
+        if (geom != null) {
+            size += (long) geom.getNumPoints() * BYTES_PER_GEOMETRY_POINT;
+        }
+        return size;
     }
 
     @Override
@@ -151,6 +187,7 @@ public class Importer implements de.komoot.photon.Importer {
         final BulkRequest req = b.builder.build();
         b.builder = new BulkRequest.Builder();
         b.todo = 0;
+        b.bytes = 0;
 
         try {
             bulkQueue.put(req);

--- a/src/test/java/de/komoot/photon/opensearch/ImporterEstimateSizeTest.java
+++ b/src/test/java/de/komoot/photon/opensearch/ImporterEstimateSizeTest.java
@@ -1,0 +1,43 @@
+package de.komoot.photon.opensearch;
+
+import de.komoot.photon.PhotonDoc;
+import org.junit.jupiter.api.Test;
+import org.locationtech.jts.io.ParseException;
+import org.locationtech.jts.io.WKTReader;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * Unit tests for the bulk-flush byte estimate. The estimate is what keeps a -full-geometry
+ * import from building a bulk request larger than OpenSearch's http.max_content_length, so the
+ * key property is that it grows with the number of geometry points.
+ */
+class ImporterEstimateSizeTest {
+
+    private static PhotonDoc doc() {
+        return new PhotonDoc("1", "N", 1, "place", "city");
+    }
+
+    @Test
+    void geometrylessDocIsEstimatedAtTheFlatBase() {
+        assertThat(Importer.estimateSize(doc())).isEqualTo(Importer.BASE_DOC_BYTES);
+    }
+
+    @Test
+    void pointGeometryCountsAsOnePoint() throws ParseException {
+        PhotonDoc doc = doc().geometry(new WKTReader().read("POINT (6.1 51.2)"));
+
+        assertThat(Importer.estimateSize(doc))
+                .isEqualTo(Importer.BASE_DOC_BYTES + Importer.BYTES_PER_GEOMETRY_POINT);
+    }
+
+    @Test
+    void estimateGrowsWithGeometryPointCount() throws ParseException {
+        // A closed triangle ring has 4 points (the first is repeated as the last).
+        PhotonDoc doc = doc().geometry(new WKTReader().read(
+                "POLYGON ((6.1 51.2, 6.2 51.2, 6.2 51.3, 6.1 51.2))"));
+
+        assertThat(Importer.estimateSize(doc))
+                .isEqualTo(Importer.BASE_DOC_BYTES + 4L * Importer.BYTES_PER_GEOMETRY_POINT);
+    }
+}

